File: concrete/src/main/java/pcb/uwu/unit/quantity/GramUnit.java
package pcb.uwu.unit.quantity;

import pcb.uwu.units.quantity.MassUnit;

import java.math.BigDecimal;

import static java.math.MathContext.UNLIMITED;

public class GramUnit extends MassUnit {

	private static final BigDecimal GRAMS_PER_KILOGRAM = new BigDecimal(1_000);

	public static final GramUnit GRAM = new GramUnit();

	private GramUnit() {
		super("g", "gram", "grams",
				amount -> amount.dividedBy(GRAMS_PER_KILOGRAM, UNLIMITED),
				amount -> amount.multipliedBy(GRAMS_PER_KILOGRAM, UNLIMITED));
	}
}

File: implementation/src/main/java/pcb/uwu/unit/base/GramUnit.kt
package pcb.uwu.unit.base

import pcb.uwu.unit.base.GramUnit.CONVERT

object GramUnit
    : MassUnit(symbol = "g",
               singularName = "gram",
               pluralName = "grams",
               translationToCanonical = { it * CONVERT },
               translationFromCanonical = { it / CONVERT })
{
    const val CONVERT = 1000
}
